lib/spack/spack/solver/asp.py

```diff
@@ -2796,6 +2823,154 @@ def _specs_from_requires(self, pkg_name, section):
                 yield _spec_with_default_name(s, pkg_name)
 
 
+class RuntimePropertyRecorder:
+    """An object of this class is injected in callbacks to compilers, to let them declare
+    properties of the runtimes they support and of the runtimes they provide, and to add
+    runtime dependencies to the nodes using said compiler.
+
+    The usage of the object is the following. First, a runtime package name or the wildcard
+    "*" are passed as an argument to __call__, to set which kind of package we are referring to.
+    Then we can call one method with a directive-like API.
```

Comment:
I think the name of this thing is too specific and it's really not *only* about runtimes.  What about `PackageConstraints` or `PackageConstraintContext` or something like that?

There are other things we may want to add global constraints for (e.g. other configuration, or flags that need to be consistent, etc.) so I could see this becoming a very general mechanism that any package could add a callback for.

Reply:
I don't think we need to rename this for the PR.  Can do later.